zio-http/src/main/scala/zio/http/service/ZKeyedPool.scala

```diff
@@ -0,0 +1,173 @@
+package zio.http.service
+
+import zio._
+import zio.stm.{TMap, TSet, ZSTM}
+
+trait ZKeyedPool[+Err, -Key, Item] {
+
+  /**
+   * Retrieves an item from the pool belonging to the given key in a scoped
+   * effect. Note that if acquisition fails, then the returned effect will fail
+   * for that same reason. Retrying a failed acquisition attempt will repeat the
+   * acquisition attempt.
+   */
+  def get(key: Key)(implicit trace: Trace): ZIO[Scope, Err, Item]
+
+  /**
+   * Invalidates the specified item. This will cause the pool to eventually
+   * reallocate the item, although this reallocation may occur lazily rather
+   * than eagerly.
+   */
+  def invalidate(item: Item)(implicit trace: Trace): UIO[Unit]
+
+}
+
+object ZKeyedPool {
+
+  /**
+   * Makes a new pool of the specified fixed size. The pool is returned in a
+   * `Scope`, which governs the lifetime of the pool. When the pool is shutdown
+   * because the `Scope` is closed, the individual items allocated by the pool
+   * will be released in some unspecified order.
+   */
+  def make[Key, Env: Tag, Err, Item](get: Key => ZIO[Env, Err, Item], size: => Int)(implicit
+    trace: Trace,
+  ): ZIO[Env with Scope, Nothing, ZKeyedPool[Err, Key, Item]] =
+    make(get, _ => size)
+
+  /**
+   * Makes a new pool of the specified fixed size. The pool is returned in a
+   * `Scope`, which governs the lifetime of the pool. When the pool is shutdown
+   * because the `Scope` is closed, the individual items allocated by the pool
+   * will be released in some unspecified order.
+   *
+   * The size of the underlying pools can be configured per key.
+   */
+  def make[Key, Env: Tag, Err, Item](get: Key => ZIO[Env, Err, Item], size: Key => Int)(implicit
+    trace: Trace,
+  ): ZIO[Env with Scope, Nothing, ZKeyedPool[Err, Key, Item]] = {
+    makeWith(get, (key: Key) => { val s = size(key); s to s })(_ => None)
+  }
+
+  /**
+   * Makes a new pool with the specified minimum and maximum sizes and time to
+   * live before a pool whose excess items are not being used will be shrunk
+   * down to the minimum size. The pool is returned in a `Scope`, which governs
+   * the lifetime of the pool. When the pool is shutdown because the `Scope` is
+   * used, the individual items allocated by the pool will be released in some
+   * unspecified order.
+   */
+  def make[Key, Env: Tag, Err, Item](get: Key => ZIO[Env, Err, Item], range: => Range, timeToLive: => Duration)(implicit
```

Comment:
As written, it allows infinitely many pools. That's like Netty although if we are using this in the HTTP client, it does mean the number of unique domains should be fixed and small (e.g. no web crawlers).

The moment you impose a cap, however, you have "cache problems", since you now need to do eviction on "unused" or "less used" resources when new requests come in.

Reply:
Yes. In the context of HTTP connection pools I was thinking that we never want to limit the number of pools - that would mean that when connecting to a new host you fall back to a "no connection pool" mode. All the other options seems to be invalid:
- you cannot block until another connection pool is closed because that normally does not happen - it keeps connections alive.
- you cannot (easily) force another one to shut down because it may have active connections 

I think when dealing with dynamic hosts, you just want to use the `ConnectionPool.disabled` and you can still provide a connection pool for the static subset of your clients.